### PR TITLE
feat: add preserve_output_dtypes config [noa/issue-52] swev-id: scikit-learn__scikit-learn-25102

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- add preserve_output_dtypes to global config and wrappers
- ensure pandas outputs can retain input dtypes when untouched
- cover dtype preservation, skip scenarios, and duplicate column handling

## Testing
- pytest sklearn/utils/tests/test_set_output.py -q

## Issues
- #52

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- pytest sklearn/utils/tests/test_set_output.py -q (19 passed, 0 failed, 0 skipped)
- lint: not run (not required for this change)

File: sklearn/utils/_set_output.py [hunk 1]
```diff
@@ -57,9 +68,25 @@ def _wrap_in_pandas_container(
...
+        resolved_columns = df.columns if columns is None else columns
+        if resolved_columns is not None:
+            try:
+                resolved_columns = list(resolved_columns)
+            except TypeError:
+                resolved_columns = list(df.columns)
+
+            if all(col in original_input.columns for col in resolved_columns):
+                original_subset = original_input.loc[:, resolved_columns]
+                if np.array_equal(df.to_numpy(), original_subset.to_numpy()):
```
Comment (noa-lucent) [inline sklearn/utils/_set_output.py]: [major] Thanks for the thorough guardrails here. One important edge case slips through:  treats  values as unequal by default. As a result, any column that contains missing values will fail this equality check and we silently opt out of dtype preservation even though the transformer returned the unmodified data. The enhancement in #52 is supposed to cover those cases too, so the new config effectively won’t work on realistic data with nulls. Please switch to an equality check that considers s equal—e.g. —and add a regression test covering a column with s so we don’t regress here.}

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for the enhancement—most of the scaffolding looks solid. I spotted one blocking edge case: the dtype preservation check uses  without , so any column containing NaNs fails the match and we silently drop back to the old behaviour. That breaks the main promise of #52 for real data with missing values. Please address the equality check (and cover it with a regression test).

File: sklearn/utils/_set_output.py (hunk 1 above)
Comment (rowan-stein) [reply sklearn/utils/_set_output.py]: Thanks, great catch. We’ll update the equality check to handle NaNs (using `np.array_equal(..., equal_nan=True)`) and add a regression test that includes missing values to ensure preservation works on realistic data.

File: sklearn/utils/tests/test_set_output.py [hunk 2]
```diff
@@ -5,6 +5,9 @@
 from numpy.testing import assert_array_equal
 
 from sklearn._config import config_context, get_config
+from sklearn.base import TransformerMixin
+from sklearn.feature_selection import SelectKBest, f_classif
```
Comment (noa-lucent) [inline sklearn/utils/tests/test_set_output.py:9]: [nit] Looks like  is no longer used in this module—could you drop the import while you’re here?

**Review by noa-lucent — APPROVED** [review]
NaN preservation looks solid now—thanks for adding the regression test and tightening the equality check. I left one small nit about an unused import, otherwise this is good to go.
